File: 프로그래머스/1/42862. 체육복/체육복.swift
// 도난

// 번호
// 앞 아니면 뒤만 빌려줄 수 있음

// 전체 n명
// 도난 lost 리스트
// 여벌 reserve 리스트

// 최대 학생수 return

// n: 2~30
// lost: 1~n 중복 x
// reserve: 1~n 중복 x
// 여벌이 도난이면 빌려줄 수 없음

// lostSet, reserveSet -> contain여부로 확인
// 1. 본인 도난 체크
// 2. 본인 여벌 체크
// 3. 주변 필요 체크

import Foundation

func solution(_ n:Int, _ lost:[Int], _ reserve:[Int]) -> Int {
    
    var stu = 0
    var lostSet = Set(lost)
    var reserveSet = Set(reserve)
    
    for i in (1...n) {
        
        let isLost = lostSet.contains(i)
        let hasReserve = reserveSet.contains(i)
        
        if isLost {
            // 도난 O
            
            if hasReserve { 
                // 여분 O
                reserveSet.remove(i) 
                stu += 1
                
            } else {
                
                // 여분 X
                let hasFront = reserveSet.contains(i-1) && !lostSet.contains(i-1)
                let hasBack = reserveSet.contains(i+1) && !lostSet.contains(i+1)
                
                if hasFront {
                    reserveSet.remove(i-1) 
                    stu += 1
                } else if hasBack {
                    reserveSet.remove(i+1) 
                    stu += 1
                }
            }
            
        } else {
            // 도난 X
            stu += 1
        }
    }
    
    return stu
}

File: 프로그래머스/1/42862. 체육복/체육복.py
# 도둑 > 체육복 도난
# 여벌 > 빌려주기

# 1 2 3 4 5 > 앞 혹은 뒤에게만 빌려줄 수 있다 
# 최대한 많은 학생이 수업

# n명의 학생
# [1,2,3] lost 배열
# [1,2,3] reserve 배열

# 수업 듣는 최대 학생수

# n: 2~30명
# 여벌 & 도난이면 빌려줄 수 없다

# [도난, 여분]
# loop돌면서
    # 앞 (도난 & 현재 X) & 현재 (도난 X & 여벌 O) 앞으로 전달

def solution(n, lost, reserve):
    
    arr = [[0, 0] for _ in range(n)]
    
    for l in lost:
        arr[l-1][0] = 1
    
    for r in reserve:
        if arr[r-1][0] == 1:
            arr[r-1][0] = 0
            continue
        arr[r-1][1] = 1
    
    for i in range(n):
        i_lost, i_reserve = arr[i]
            
        # 앞 사람 체크 (lost이면)
        if i and arr[i-1][0] and i_reserve:
            arr[i-1][0] = 0
            arr[i][1] = 0
            continue
        
        # 뒷 사람 체크
        if i+1 < n and arr[i+1][0] and i_reserve:
            arr[i+1][0] = 0
            arr[i][1] = 0
            continue
            
    return n - sum(list(map(lambda x: x[0], arr)))

# 19분 소요